python/ray/data/tests/test_block_sizing.py

```diff
@@ -4,36 +4,109 @@
+)
 from ray.tests.conftest import *  # noqa
 
 
-def test_map(ray_start_regular_shared, restore_data_context):
+def _assert_plasma_metrics(cursor, num_blocks_expected):
     ctx = DataContext.get_current()
-    ctx.target_min_block_size = 100 * 8
-
-    # NOTE(swang): For some reason BlockBuilder's estimated
-    # memory usage is 2x the actual memory usage.
-    num_blocks_expected = 20
-    ctx.target_max_block_size = 100 * 8
 
-    ds = ray.data.range(1000).map(lambda row: row)
-    assert ds.materialize().num_blocks() == num_blocks_expected
-    ds = ray.data.range(1000).map(lambda row: row).map(lambda row: row)
-    assert ds.materialize().num_blocks() == num_blocks_expected
+    try:
+        cursor = assert_core_execution_metrics_equals(
+            CoreExecutionMetrics(
+                object_store_stats={
+                    "cumulative_created_plasma_objects": lambda count: count
+                    >= num_blocks_expected
+                    and count < 2 * num_blocks_expected,
+                    "cumulative_created_plasma_bytes": lambda count: count
+                    >= ctx.target_max_block_size * num_blocks_expected
+                    and count < ctx.target_max_block_size * (num_blocks_expected + 1),
+                },
+            ),
+            cursor,
+        )
+    except AssertionError:
+        return False
+    return True
+
+
+def test_map(shutdown_only, restore_data_context):
+    ray.init(
+        _system_config={
+            "max_direct_call_object_size": 10_000,
+        },
+        num_cpus=2,
+        object_store_memory=int(100e6),
+    )
 
+    ctx = DataContext.get_current()
+    ctx.target_min_block_size = 10_000 * 8
+    ctx.target_max_block_size = 10_000 * 8
+    num_blocks_expected = 10
+    cursor = get_initial_core_execution_metrics_cursor()
+
+    # Test read.
+    ds = ray.data.range(100_000, parallelism=1).materialize()
+    assert num_blocks_expected <= ds.num_blocks() <= num_blocks_expected + 1
+    cursor = assert_blocks_expected_in_plasma(
+        cursor, num_blocks_expected, block_size_expected=ctx.target_max_block_size
+    )
+
+    # Test read -> map.
+    # NOTE(swang): For some reason BlockBuilder's estimated memory usage when a
+    # map fn is used is 2x the actual memory usage.
+    ds = ray.data.range(100_000, parallelism=1).map(lambda row: row).materialize()
+    assert num_blocks_expected * 2 <= ds.num_blocks() <= num_blocks_expected * 2 + 1
+    cursor = assert_blocks_expected_in_plasma(
+        cursor,
+        num_blocks_expected * 2,
+        block_size_expected=ctx.target_max_block_size // 2,
+    )
+
+    # Test adjusted block size.
     ctx.target_max_block_size *= 2
-    ds = ray.data.range(1000).map(lambda row: row)
-    assert ds.materialize().num_blocks() * 2 == num_blocks_expected
-    ds = ray.data.range(1000).map(lambda row: row).map(lambda row: row)
-    assert ds.materialize().num_blocks() * 2 == num_blocks_expected
+    num_blocks_expected /= 2
```

Comment:
```suggestion
    num_blocks_expected //= 2
```
In `equals_or_true`, we check "type = int". So make sure not accidentally convert int to float. 
Or maybe we should change the condition to `numbers.Number` there.